rig/routing_table/entries.py

```diff
@@ -23,7 +24,18 @@ class RoutingTableEntry(namedtuple("RoutingTableEntry", "route key mask")):
     mask : int
         32-bit unsigned integer mask to apply to keys of packets arriving at
         the router.
+    sources : {:py:class:`~.Routes`, ...}
+        Directions in which a packet may enter the router before taking this
+        route. This is opposite to the link used (e.g., ``north`` in this set
+        indicates that a packet would enter on the ``south`` link). If the
+        `sources` and `route` are equivalent and contain only one item then it
+        *might* be possible to remove the entry and instead use default
+        routing.
     """
+    def __new__(cls, route, key, mask, sources={None}):
+        return super(RoutingTableEntry, cls).__new__(
+                cls, route, key, mask, set(sources))
+
     def __str__(self):
         """Get an easily readable representation of the routing table entry.
```

Comment:
Tricky... Gut feeling is probably this should be the name of the link the packet arrived through (not the direction of travel). After all, "Core 1" cannot be "inverted" and if something comes on on North, and West, say, it would be confusing to invert that. Thoughts?